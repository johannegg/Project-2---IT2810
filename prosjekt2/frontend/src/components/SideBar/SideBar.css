.sidebar {
	position: fixed;
	top: 80px;
	bottom: 0;
	left: 0;
	background-color: var(--theme-pink);
	width: 300px;
	height: calc(100vh - 130px); /* looks a bit weird, but it works and we can fix later;)*/
	padding-top: 60px;
	box-shadow: 2px 0 5px rgba(0, 0, 0, 0.2);
	transform: translateX(-100%);
	transition: transform 0.3s ease;
	z-index: 998;
	overflow-y: auto;
}

.sidebar.open {
	transform: translateX(0);
}

.close-button {
	position: absolute;
	top: 5px;
	right: 10px;
	background: none;
	border: none;
	font-size: 32px;
	cursor: pointer;
	color: white;
	z-index: 1000;
}

.close-button:hover {
	color: var(--theme-light-pink);
}

.filteringContainer {
	display: flex;
	flex-direction: column;
	gap: 5px;
	padding: 20px;
	align-items: center;
	width: 90%;
}

.clearFiltersButton {
    padding: 7px;
    border: none;
    border-radius: 8px; 
    cursor: pointer;
    font-size: 16px; 
	background-color: white;
}


@media (max-width: 1050px) {
	.sidebar {
		width: 200px;
	}
}

@media (max-width: 600px) {
	.sidebar {
		width: 85%;
	}
}

@media (max-width: 500px) {
	.hamburger-icon {
		top: 10px;
		left: 5px;
	}
}

@media (prefers-color-scheme: dark) {
	.sidebar {
		background-color: var(--theme-even-darker-pink);
	}

	.clearFiltersButton  {
		background-color: var(--theme-almost-black);
	}
}
