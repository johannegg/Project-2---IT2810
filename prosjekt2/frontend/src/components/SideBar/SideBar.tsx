import "./Sidebar.css";
import { Filter } from "../GenreFilter/GenreFilter";
import { ViewsFilter } from "../ViewsFilter/ViewsFilter";
import Sort from "../Sort/Sort";
import { SongData } from "../../utils/types/SongTypes";

type SidebarProps = {
	onGenreChange: (selectedGenres: string[]) => void;
	onViewsChange: (minViews: number, maxViews: number) => void;
	sortOption: string;
	onSortChange: (newSort: string) => void;
	songs: SongData[];
	onToggle: (isOpen: boolean) => void;
	isOpen: boolean;
};

export function Sidebar({
	onGenreChange,
	onViewsChange,
	sortOption,
	onSortChange,
	songs,
	onToggle,
	isOpen,
}: SidebarProps) {
	const toggleMenu = () => {
		onToggle(!isOpen);
	};

	return (
		<div className={`sidebar ${isOpen ? "open" : ""}`}>
			<button className="close-button" onClick={toggleMenu} type="button">
				âœ•
			</button>
			<div className="filteringContainer">
				<Filter onGenreChange={onGenreChange} songs={songs} />
				<br />
				<ViewsFilter onViewsChange={onViewsChange} />
				<br />
				<Sort songs={songs} sortOption={sortOption} onSortChange={onSortChange} />
			</div>
		</div>
	);
}
