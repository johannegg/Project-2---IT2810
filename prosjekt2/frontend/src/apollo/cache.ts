import { InMemoryCache, makeVar } from "@apollo/client";
import { SongData } from "../utils/types/SongTypes";
import { PlaylistData } from "../pages/Playlists/Playlists";

// Initialize sessionStorage and localStorage if values are not already set
const initializeStorage = (key: string, defaultValue: string) => {
  if (!sessionStorage.getItem(key)) {
    sessionStorage.setItem(key, defaultValue);
  }
};

const initializeLocalStorage = (key: string, defaultValue: string) => {
  if (!localStorage.getItem(key)) {
    localStorage.setItem(key, defaultValue);
  }
};

// Initialize values in storage
initializeStorage("minViews", "0");
initializeStorage("maxViews", "1000000");
initializeStorage("selectedGenres", "[]");
initializeStorage("sortOption", "views_desc");
initializeStorage("homeSearchTerm", "");
initializeStorage("favoritesSearchTerm", "");
initializeLocalStorage("favoriteSongs", "[]");
initializeLocalStorage("playlists", "[]");

// Load saved values from storage
const savedGenres = JSON.parse(sessionStorage.getItem("selectedGenres") || "[]");
const savedMinViews = Number(sessionStorage.getItem("minViews"));
const savedMaxViews = Number(sessionStorage.getItem("maxViews"));
const savedSortOption = sessionStorage.getItem("sortOption") || "views_desc";
const savedHomeSearchTerm = sessionStorage.getItem("homeSearchTerm") || "";
const savedFavoritesSearchTerm = sessionStorage.getItem("favoritesSearchTerm") || "";
const savedFavoriteSongs = JSON.parse(localStorage.getItem("favoriteSongs") || "[]");
const savedPlaylists = JSON.parse(localStorage.getItem("playlists") || "[]");

// Create reactive variables
export const genreFilterVar = makeVar<string[]>(savedGenres);
export const sortOptionVar = makeVar<string>(savedSortOption);
export const minViewsVar = makeVar<number>(savedMinViews);
export const maxViewsVar = makeVar<number>(savedMaxViews);
export const homeSearchTermVar = makeVar<string>(savedHomeSearchTerm);
export const favoritesSearchTermVar = makeVar<string>(savedFavoritesSearchTerm);
export const favoriteSongsVar = makeVar<SongData[]>(savedFavoriteSongs);
export const playlistsVar = makeVar<PlaylistData[]>(savedPlaylists);  
export const songDataVar = makeVar<SongData[]>([]);

// Configure cache with reactive variables
const cache = new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        genreFilter: {
          read() {
            return genreFilterVar();
          },
        },
        sortOption: {
          read() {
            return sortOptionVar();
          },
        },
        minViews: {
          read() {
            return minViewsVar();
          },
        },
        maxViews: {
          read() {
            return maxViewsVar();
          },
        },
        homeSearchTerm: {
          read() {
            return homeSearchTermVar();
          },
        },
        favoritesSearchTerm: {
          read() {
            return favoritesSearchTermVar();
          },
        },
        favoriteSongs: {
          read() {
            return favoriteSongsVar();
          },
        },
        playlists: {
          read() {
            return playlistsVar();
          },
        },
        songData: {
          read() {
            return songDataVar();
          },
        },
      },
    },
  },
});

// Sync reactive variables with sessionStorage and localStorage
homeSearchTermVar.onNextChange((newHomeSearchTerm) => {
  sessionStorage.setItem("homeSearchTerm", newHomeSearchTerm);
});

favoritesSearchTermVar.onNextChange((newFavoritesSearchTerm) => {
  sessionStorage.setItem("favoritesSearchTerm", newFavoritesSearchTerm);
});

genreFilterVar.onNextChange((newGenres) => {
  sessionStorage.setItem("selectedGenres", JSON.stringify(newGenres));
});

minViewsVar.onNextChange((newMinViews) => {
  sessionStorage.setItem("minViews", String(newMinViews));
});

maxViewsVar.onNextChange((newMaxViews) => {
  sessionStorage.setItem("maxViews", String(newMaxViews));
});

sortOptionVar.onNextChange((newSortOption) => {
  sessionStorage.setItem("sortOption", newSortOption);
});

favoriteSongsVar.onNextChange((newFavorites) => {
  localStorage.setItem("favoriteSongs", JSON.stringify(newFavorites));
});

playlistsVar.onNextChange((newPlaylists) => {
  localStorage.setItem("playlists", JSON.stringify(newPlaylists));
  window.dispatchEvent(new Event("storage")); 
});

export default cache;
