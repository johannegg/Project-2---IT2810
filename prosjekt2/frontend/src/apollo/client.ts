import { ApolloClient } from "@apollo/client";
import cache from "./cache";

const client = new ApolloClient({
	uri: "http://localhost:3001/", // Use this to run locally, for tests
	//uri: "http://it2810-12.idi.ntnu.no:3001", // Use this to run on virtual machine
	cache: cache,
});

export default client;
