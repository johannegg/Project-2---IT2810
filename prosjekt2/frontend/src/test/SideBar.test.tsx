import { describe, test, vi, beforeEach, expect } from "vitest";
import { render, screen, fireEvent, cleanup, act } from "@testing-library/react";

import { isSidebarOpenVar, clearFiltersVar } from "../apollo/cache";
import { Sidebar } from "../components/SideBar/SideBar";

// Mock dependencies for child components to simplify testing
vi.mock("../components/GenreFilter/GenreFilter", () => ({
	Filter: ({ onGenreChange }: { onGenreChange: (genres: string[]) => void }) => (
		<div data-testid="genre-filter" onClick={() => onGenreChange(["Pop"])}>
			Genre Filter
		</div>
	),
}));
vi.mock("../components/ViewsFilter/ViewsFilter", () => ({
	ViewsFilter: ({ onViewsChange }: { onViewsChange: (min: number, max: number) => void }) => (
		<div data-testid="views-filter" onClick={() => onViewsChange(100, 1000)}>
			Views Filter
		</div>
	),
}));
vi.mock("../components/Sort/Sort", () => ({
	default: ({ onSortChange }: { onSortChange: (newSort: string) => void }) => (
		<div data-testid="sort" onClick={() => onSortChange("title_asc")}>
			Sort
		</div>
	),
}));

// Describe the test suite for the Sidebar component
describe("Sidebar", () => {
	// Mock functions to simulate the behavior of the props
	const mockOnGenreChange = vi.fn();
	const mockOnViewsChange = vi.fn();
	const mockOnSortChange = vi.fn();
	const mockOnToggle = vi.fn();
	const mockOnClearAllFilters = vi.fn();

	// Before each test, reset all mocks, variables, and clean up the DOM
	beforeEach(() => {
		vi.resetAllMocks();
		isSidebarOpenVar(false);
		clearFiltersVar(false);
		cleanup();
	});

	// Snapshot Test: Verifies the default rendering of Sidebar in a closed state
	test("matches snapshot when Sidebar is closed by default", () => {
		const { asFragment } = render(
			<Sidebar
				onGenreChange={mockOnGenreChange}
				onViewsChange={mockOnViewsChange}
				onSortChange={mockOnSortChange}
				onToggle={mockOnToggle}
				onClearAllFilters={mockOnClearAllFilters}
				searchTerm=""
				songs={[]}
			/>,
		);

		// Ensure that the rendered Sidebar matches the stored snapshot
		expect(asFragment()).toMatchSnapshot();
	});

	// Snapshot Test: Verifies the rendering of Sidebar when it is open
	test("matches snapshot when Sidebar is open", () => {
		isSidebarOpenVar(true);

		const { asFragment } = render(
			<Sidebar
				onGenreChange={mockOnGenreChange}
				onViewsChange={mockOnViewsChange}
				onSortChange={mockOnSortChange}
				onToggle={mockOnToggle}
				onClearAllFilters={mockOnClearAllFilters}
				searchTerm=""
				songs={[]}
			/>,
		);

		// Ensure that the rendered Sidebar matches the stored snapshot
		expect(asFragment()).toMatchSnapshot();
	});

	// Test: Verifies the Sidebar renders in a closed state by default
	test("renders Sidebar in closed state by default", () => {
		render(
			<Sidebar
				onGenreChange={mockOnGenreChange}
				onViewsChange={mockOnViewsChange}
				onSortChange={mockOnSortChange}
				onToggle={mockOnToggle}
				onClearAllFilters={mockOnClearAllFilters}
				searchTerm=""
				songs={[]}
			/>,
		);

		// Select the Sidebar element and verify its initial class states
		const sidebar = screen.getByRole("complementary", { hidden: true });
		expect(sidebar).toHaveClass("sidebar");
		expect(sidebar).not.toHaveClass("open");
	});

	// Test: Verifies the Sidebar toggles between open and closed states
	test("toggles Sidebar open and closed", () => {
		render(
			<Sidebar
				onGenreChange={mockOnGenreChange}
				onViewsChange={mockOnViewsChange}
				onSortChange={mockOnSortChange}
				onToggle={mockOnToggle}
				onClearAllFilters={mockOnClearAllFilters}
				searchTerm=""
				songs={[]}
			/>,
		);

		// Find the Close button and simulate a click to open the Sidebar
		const closeButton = screen.getByRole("button", { name: /close/i, hidden: true });
		act(() => fireEvent.click(closeButton));

		// Assert the Sidebar state is now open and the onToggle callback was called
		expect(isSidebarOpenVar()).toBe(true);
		expect(mockOnToggle).toHaveBeenCalledWith(true);

		// Simulate another click to close the Sidebar
		act(() => fireEvent.click(closeButton));

		// Assert the Sidebar state is now closed and the onToggle callback was called
		expect(isSidebarOpenVar()).toBe(false);
		expect(mockOnToggle).toHaveBeenCalledWith(false);
	});

	// Test: Verifies the Clear Filters button clears all filters
	test("clears all filters when Clear Filters button is clicked", () => {
		render(
			<Sidebar
				onGenreChange={mockOnGenreChange}
				onViewsChange={mockOnViewsChange}
				onSortChange={mockOnSortChange}
				onToggle={mockOnToggle}
				onClearAllFilters={mockOnClearAllFilters}
				searchTerm=""
				songs={[]}
			/>,
		);

		// Find the Clear Filters button and simulate a click
		const clearFiltersButton = screen.getByRole("button", { name: /clear all filters/i, hidden: true });
		fireEvent.click(clearFiltersButton);

		// Assert the onClearAllFilters callback was called
		expect(mockOnClearAllFilters).toHaveBeenCalled();

		// Assert the clearFiltersVar reactive variable state was reset
		expect(clearFiltersVar()).toBe(false);
	});
});
