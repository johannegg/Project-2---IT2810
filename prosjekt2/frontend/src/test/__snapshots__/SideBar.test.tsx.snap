// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar > matches snapshot when Sidebar is closed by default 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="sidebar "
    role="complementary"
    tabindex="-1"
  >
    <button
      aria-label="Close"
      class="close-button"
      tabindex="-1"
      type="button"
    >
      ✕
    </button>
    <div
      class="filteringContainer"
    >
      <div
        data-testid="sort"
      >
        Sort
      </div>
      <br />
      <div
        data-testid="genre-filter"
      >
        Genre Filter
      </div>
      <br />
      <div
        data-testid="views-filter"
      >
        Views Filter
      </div>
      <br />
      <button
        class="clearFiltersButton"
        tabindex="-1"
      >
        Clear filters
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Sidebar > matches snapshot when Sidebar is open 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="sidebar "
    role="complementary"
    tabindex="-1"
  >
    <button
      aria-label="Close"
      class="close-button"
      tabindex="-1"
      type="button"
    >
      ✕
    </button>
    <div
      class="filteringContainer"
    >
      <div
        data-testid="sort"
      >
        Sort
      </div>
      <br />
      <div
        data-testid="genre-filter"
      >
        Genre Filter
      </div>
      <br />
      <div
        data-testid="views-filter"
      >
        Views Filter
      </div>
      <br />
      <button
        class="clearFiltersButton"
        tabindex="-1"
      >
        Clear filters
      </button>
    </div>
  </div>
</DocumentFragment>
`;
